bokeh/charts/data_source.py

```diff
@@ -214,22 +298,49 @@ def get_selections(self, selections, **kwargs):
         return select_map
 
     def apply_operations(self):
+        """Applies each data operation."""
         # ToDo: Handle order of operation application, see GoG pg. 71
         for dim, select in iteritems(self._selections):
             if isinstance(select, DataOperator):
                 self._data = select.apply(self)
 
+    def setup_derived_columns(self):
+        """Attempt to add special case columns to the DataFrame for the builder."""
+        for dim in self._dims:
+            dim_selection = self[dim]
+            if dim_selection is not None and isinstance(dim_selection, str) and \
+                dim_selection in special_columns and dim_selection not in \
+                    self.df.columns.tolist():
+                        self._data[dim_selection] = special_columns[dim_selection](
+                            self._data)
+
     def __getitem__(self, dim):
         """Get the columns selected for the given dimension name.
 
+        If the dimension is not in `_selections`, `None` is returned.
+
         e.g. dim='x'
+
+        Returns:
+            Str or List(Str): the columns selected
         """
         if dim in self._selections:
             return self._selections[dim]
         else:
             return None
 
-    def stack_measures(self, measures, ids=None, var_name='variable', value_name='value'):
+    def __setitem__(self, dim, value):
+        self._selections[dim] = value
+        self.setup_derived_columns()
+
+    def stack_measures(self, measures, ids=None, var_name='variable',
+                       value_name='value'):
+        """Converts a 'wide' table to a 'tall' table.
+
+        A wide table is one where the column names represent a categorical variable
+        and each contains only the values associated with each unique value of the
+        categorical variable.
```

Comment:
IMO this docstring is a bit misleading.. because it only talks about the table wide->tall conversion and not about melting the data with the current datagroup _data...

Also a simple example would be useful